### PR TITLE
Advanced physical inventory

### PR DESCRIPTION
**Pull request by clonedagain**
The modules added here let users handle big and complex stock inventory.
This is quite a big contribution, so to ease the review I've split it into 2 commits: 
- the 1st one is about making sure the inventory is reliable even on big installations: purging products not recorded, blocking locations etc.
- The second one is about splitting big inventories into several smaller ones, to let managers delegate part of the inventory and to get a global view of the progress during the physical counting operations.
  We have taken great care to overlap as little as possible with the new WMS in v8, so we expect the future migration to be rather easy.

Both modules have been deployed in production by at least Numérigrape and Acsone for large and complex warehouses.

This pull request supersedes the previous merge proposals made on Launchpad:
- https://code.launchpad.net/~numerigraphe-team/stock-logistic-warehouse/7.0-inventory-location/+merge/223880
- https://code.launchpad.net/~numerigraphe-team/stock-logistic-warehouse/7.0-inventory-hierarchical/+merge/223882
- https://code.launchpad.net/~acsone-openerp/stock-logistic-warehouse/some-fixes-lga/+merge/227680

### CODE REVIEW

**Comment by clonedagain** [comment]
@laetitia-gangloff can you please review this and check everything is all right and I got the latest version of your fixes please?

**Comment by laetitia-gangloff** [comment]
Hello,

the source are up to date.

:+1:

**Comment by agb80** [comment]
@clonedagain this line is including in the stock_inventory_location module description:
- every good that is not in the Inventory Lines is considered lost, and gets moved out of the stock when you confirm the Inventory.

Installed the module and tried specially that feature but didn't work. I made a code review and couldn't find any function that implement the feature on this merge request. The question is, do you have another branch where this feature is implemented? The purpose is prevent duplicate efforts, if you like we can develop this feature and propose merge.

**Comment by laetitia-gangloff** [comment]
@agb80 When you confirm inventory, lines with a quantity to 0 are created for all goods not in the inventory (and in location concerned). But to get this function, you need to check exhaustive check box.

You can see this part in this file :
wizard/stock_confirm_uninventoried_location.py:confirm_uninventoried_locations:47

Do you check exhaustive checkbox ?
Do you get this wizard open, when you confirm your inventory ?

**Comment by clonedagain** [comment]
@agb80 we implemented it the way @laetitia-gangloff described to minimize the code impact, knowing that v8 has a whole new WMS.

**Comment by sbidoul** [comment]
@clonedagain have you seen the flake8 errors?

**Comment by clonedagain** [comment]
I'll check it sorry. @laetitia-gangloff has alredy done changes for this, I'll see if I've merged them all right

**Comment by clonedagain** [comment]
OK I had actually forgotten @laetitia-gangloff's patch but now it's merged, with 3 small fixes more.
Rebased on 7.0, should be all right now.

**Comment by clonedagain** [comment]
I'm not sure what to do with the pylint errors that make Travis unhappy. Are we really supposed to make those changes?

**Comment by clonedagain** [comment]
Can someone please mark this "NEEDS REVIEW" ?

**Comment by clonedagain** [comment]
@jgrandguillaume I've read you have logistics projects going on, would you care to give these modules a try and tell us what you think please?
If they are accepted, we can start migrating them to v8.

**Comment by sbidoul** [comment]
@clonedagain the pylint errors should be fixed by manually cleaning up #20 then rebasing.

**Comment by lepistone** [comment]
The 7.0 branch is green because lint problems were sorted in #21. I closed #20 which is now unnecessary.

Rebasing this one on 7.0 should do!

**Comment by jgrandguillaume** [comment]
Hi,

I asked to have the runbot on this project so we'll be able to test this PR from a functionnal point of view.

Regards,

**Comment by clonedagain** [comment]
Fantastic, keep me posted!
@lbellier will start porting this to v8 before the summer, so we'd be happy to have as much feedback as possible.

**Comment by yvaucher** [comment]
Could those commits be linked to @lbellier account?

**Comment by clonedagain** [comment]
@yvaucher  we'll look into it, they should indeed be linked to his github account.

**Comment by clonedagain** [comment]
Rebased and fixed the wrong email address of Loïc.
I've signed the CLA too, can you please update the CLAbot database?

File: stock_inventory_hierarchical/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+{
+    "name": "Hierarchical Physical Inventory",
+    "version": "1.1",
+    "depends": ["stock"],
+    "author": "Numérigraphe",
```
Comment (yvaucher) [inline stock_inventory_hierarchical/__openerp__.py]: Please add Odoo Community Association (OCA) under author

File: stock_inventory_hierarchical/i18n/fr.po [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+#. module: stock_inventory_hierarchical
+#: field:stock.inventory,complete_name:0
+msgid "Complete reference"
+msgstr "Réference complète"
+
+#. module: stock_inventory_hierarchical
+#: constraint:stock.inventory:0
+msgid "Error: You can not create recursive inventories."
+msgstr "Erreur : Vous ne pouvez pas créer d'inventaire récursifs."
```
Comment (yvaucher) [inline stock_inventory_hierarchical/i18n/fr.po]: s/récursifs/récursif

File: stock_inventory_hierarchical/test/hierarchical_inventory_test.yml [hunk 3]
```diff
@@ -0,0 +1,96 @@
+-
+  In this file, i check rules about hierarchical inventories.
```
Comment (yvaucher) [inline stock_inventory_hierarchical/test/hierarchical_inventory_test.yml]: s/i check/I check

File: stock_inventory_hierarchical/test/hierarchical_inventory_test.yml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    from stock_inventory_hierarchical import HierarchicalInventoryException
+    try:
+      self.action_done(cr, uid, [ref('stock_inventory_parent0')])
+    except HierarchicalInventoryException as e:
+      log("Good, the inventory could not be validated: %s", e)
+    parent_state = self.read(cr, uid, [ref('stock_inventory_parent0')], ['state'])[0]['state']
+    assert parent_state == 'confirm', "Parent inventory have '%s' state. It should be 'confirm'" % parent_state
+
+-
+  Now, i'll validate all children inventory before validate the parent.
```
Comment (yvaucher) [inline stock_inventory_hierarchical/test/hierarchical_inventory_test.yml]: s/i'll/I'll

File: stock_inventory_hierarchical_location/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+{
+    "name": "Exhaustive and hierarchical Stock Inventories",
+    "version": "1.1",
+    "author": u"Numérigraphe",
```
Comment (yvaucher) [inline stock_inventory_hierarchical_location/__openerp__.py]: Please add OCA in author list

File: stock_inventory_hierarchical_location/__openerp__.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+{
+    "name": "Exhaustive and hierarchical Stock Inventories",
```
Comment (yvaucher) [inline stock_inventory_hierarchical_location/__openerp__.py:22]: A summary would be great

File: stock_inventory_hierarchical/__openerp__.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+{
+    "name": "Hierarchical Physical Inventory",
```
Comment (yvaucher) [inline stock_inventory_hierarchical/__openerp__.py:22]: A summary would be great

**Comment by yvaucher** [comment]
@clonedagain Thanks commits are fixed I just found some little things to fix and it's :+1:

File: stock_inventory_hierarchical_location/__openerp__.py (hunk 6 above)
Comment (clonedagain) [inline stock_inventory_hierarchical_location/__openerp__.py:22]: What sort of summary? do you have an example we could follow please?

Comment (yvaucher) [inline stock_inventory_hierarchical_location/__openerp__.py:22]: In `__openerp__.py` you can use the dict key `summary` as in https://github.com/OCA/vertical-ngo/blob/8.0/logistic_order_donation/__openerp__.py#L23

It is just a short description which is shown on apps and in module list which helps the user to figure out what does the module. Many OCA modules are missing it but I think we should use it as it is quite useful.

Here you see some modules with a summary bellow the title (though many are missing it)
https://www.odoo.com/apps/modules?series=8.0&order=Downloads&author=Odoo+Community+Association+%28OCA%29

**Comment by yvaucher** [comment]
Thanks :+1:

**Comment by clonedagain** [comment]
Travis runs green at last ! :-)
Runbot is unhappy still, but I'm not sure why?

**Comment by yvaucher** [comment]
@clonedagain Don't mind about runbot it has still some issues. @gurneyalex is on it

**Comment by clonedagain** [comment]
I think this reached the quorum 2 weeks ago, can it be merged?

**Comment by yvaucher** [comment]
Here we go

**Comment by clonedagain** [comment]
Thanks! Just in time for us to start migrating to v8.

**Pull request merged**